### PR TITLE
feat(core): remember the scroll position of doc when routing forward and backward

### PR DESCRIPTION
**Pull request by JimmFly**
close AF-1011

https://github.com/user-attachments/assets/d2dfeee2-926f-4760-b3fb-8baf5ff90aa9

### CODE REVIEW

**Comment by JimmFly** [comment]
* **#8631** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8631?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @JimmFly and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/frontend/core/src/desktop/pages/workspace/detail-page/detail-page.tsx [hunk 1]
```diff
@@ -233,13 +233,39 @@ const DetailPageImpl = memo(function DetailPageImpl() {
...
+  const [hasScrollTop, setHasScrollTop] = useState(false);
 
   const openOutlinePanel = useCallback(() => {
     workbench.openSidebar();
     view.activeSidebarTab('outline');
   }, [workbench, view]);
 
+  const scrollViewportRef = useRef<HTMLDivElement | null>(null);
+  const [scrolled, setScrolled] = useState(false);
+  const scrollPosition = view.scrollPositions.get(view.history.location);
```
Comment (pengx17) [inline packages/frontend/core/src/desktop/pages/workspace/detail-page/detail-page.tsx]: Move this inside handlers, eg.
 
```ts
  const handleScroll = useCallback(
    (e: React.UIEvent<HTMLDivElement>) => {
      const scrollTop = e.currentTarget.scrollTop;
      const scrollPosition = view.getScrollPosition(view.history.location);

      const hasScrollTop = scrollTop > 0;
      setHasScrollTop(hasScrollTop);
      if (scrollPosition !== scrollTop && scrolled) {
        view.setScrollPosition(scrollTop);
      }
    },
    [scrolled, view]
  );
```

**Review by pengx17 — CHANGES_REQUESTED** [review]

File: packages/frontend/core/src/desktop/pages/workspace/detail-page/detail-page.tsx [hunk 2]
```diff
@@ -233,13 +233,39 @@ const DetailPageImpl = memo(function DetailPageImpl() {
...
-  const [refCallback, hasScrollTop] = useHasScrollTop();
+  const [hasScrollTop, setHasScrollTop] = useState(false);
 
   const openOutlinePanel = useCallback(() => {
     workbench.openSidebar();
     view.activeSidebarTab('outline');
   }, [workbench, view]);
 
+  const scrollViewportRef = useRef<HTMLDivElement | null>(null);
+  const [scrolled, setScrolled] = useState(false);
```
Comment (pengx17) [inline packages/frontend/core/src/desktop/pages/workspace/detail-page/detail-page.tsx]: change to useRef. scrolled value changes may not retrigger the useEffect

File: packages/frontend/core/src/modules/workbench/view/browser-adapter.ts [hunk 3]
```diff
@@ -67,7 +61,25 @@ export function useBindWorkbenchToBrowserRouter(
...
+        );
+        if (target !== -1) {
+          const now = view.history.index;
+          const delta = target - now;
+          view.history.go(delta);
+          return;
+        }
+      }
+    }
     view.history.push(newLocation, 'fromBrowser');
```
Comment (pengx17) [inline packages/frontend/core/src/modules/workbench/view/browser-adapter.ts:83]: skip if from view?

**Review by pengx17 — APPROVED** [review]

**Review by pengx17 — COMMENTED** [review]
[clap](https://app.graphite.dev//settings/meme-library?org=toeverything)

**Pull request merged**